File: flutter_client/lib/modules/video_player/subtitle/webvtt_parser.dart
class SubtitleCue {
  final Duration start;
  final Duration end;
  final String text;

  const SubtitleCue({
    required this.start,
    required this.end,
    required this.text,
  });
}

Duration? _parseVttTimestamp(String raw) {
  final s = raw.trim();
  if (s.isEmpty) return null;
  final parts = s.split(':');
  if (parts.length < 2 || parts.length > 3) return null;
  int hours = 0;
  int minutes = 0;
  double seconds = 0;
  if (parts.length == 3) {
    hours = int.tryParse(parts[0]) ?? 0;
    minutes = int.tryParse(parts[1]) ?? 0;
    seconds = double.tryParse(parts[2].replaceAll(',', '.')) ?? 0;
  } else {
    minutes = int.tryParse(parts[0]) ?? 0;
    seconds = double.tryParse(parts[1].replaceAll(',', '.')) ?? 0;
  }
  if (hours < 0 || minutes < 0 || seconds < 0) return null;
  final totalMs = (hours * 3600 * 1000) +
      (minutes * 60 * 1000) +
      (seconds * 1000).round();
  return Duration(milliseconds: totalMs);
}

List<SubtitleCue> parseWebVtt(String vttText) {
  final lines = vttText.replaceAll('\r\n', '\n').replaceAll('\r', '\n').split('\n');
  final cues = <SubtitleCue>[];

  int i = 0;
  // Skip BOM and header
  while (i < lines.length && lines[i].trim().isEmpty) {
    i++;
  }
  if (i < lines.length) {
    final first = lines[i].trimLeft();
    if (first.startsWith('\uFEFF')) {
      lines[i] = first.substring(1);
    }
  }
  if (i < lines.length && lines[i].trim().toUpperCase().startsWith('WEBVTT')) {
    i++;
  }

  while (i < lines.length) {
    // Skip empty lines
    while (i < lines.length && lines[i].trim().isEmpty) {
      i++;
    }
    if (i >= lines.length) break;

    // Skip NOTE/STYLE/REGION blocks
    final head = lines[i].trim();
    if (head.startsWith('NOTE') || head.startsWith('STYLE') || head.startsWith('REGION')) {
      i++;
      while (i < lines.length && lines[i].trim().isNotEmpty) {
        i++;
      }
      continue;
    }

    // Optional cue identifier
    String timeLine = lines[i];
    if (!timeLine.contains('-->')) {
      i++;
      if (i >= lines.length) break;
      timeLine = lines[i];
    }

    final arrow = timeLine.indexOf('-->');
    if (arrow == -1) {
      i++;
      continue;
    }
    final left = timeLine.substring(0, arrow).trim();
    final rightPart = timeLine.substring(arrow + 3).trim();
    final right = rightPart.split(RegExp(r'\s+'))[0].trim();

    final start = _parseVttTimestamp(left);
    final end = _parseVttTimestamp(right);
    i++;
    if (start == null || end == null) {
      // skip cue text lines until blank
      while (i < lines.length && lines[i].trim().isNotEmpty) {
        i++;
      }
      continue;
    }

    final buf = StringBuffer();
    while (i < lines.length && lines[i].trim().isNotEmpty) {
      if (buf.isNotEmpty) buf.writeln();
      buf.write(lines[i]);
      i++;
    }
    final text = _formatCueText(buf.toString().trim());
    if (text.isNotEmpty && end > start) {
      cues.add(SubtitleCue(start: start, end: end, text: text));
    }
  }

  cues.sort((a, b) => a.start.compareTo(b.start));
  return cues;
}

/// Find active cue using binary search. `cues` must be sorted by start time.
SubtitleCue? findActiveCue(List<SubtitleCue> cues, Duration position) {
  if (cues.isEmpty) return null;
  int lo = 0;
  int hi = cues.length - 1;
  int best = -1;
  while (lo <= hi) {
    final mid = (lo + hi) >> 1;
    final c = cues[mid];
    if (c.start <= position) {
      best = mid;
      lo = mid + 1;
    } else {
      hi = mid - 1;
    }
  }
  if (best < 0) return null;
  final c = cues[best];
  if (position >= c.start && position <= c.end) return c;
  return null;
}

/// ffmpeg WebVTT 常带 `<br>`、`<font>` 等标签；客户端叠层需转成纯文本。
String _formatCueText(String raw) {
  var s = raw.trim();
  if (s.isEmpty) return s;
  s = s.replaceAll(RegExp(r'<br\s*/?>', caseSensitive: false), '\n');
  s = s.replaceAll(RegExp(r'<[^>]*>'), '');
  s = s.replaceAll(RegExp(r'\{[^}]*\}'), '');
  s = s
      .replaceAll('&nbsp;', ' ')
      .replaceAll('&amp;', '&')
      .replaceAll('&lt;', '<')
      .replaceAll('&gt;', '>')
      .replaceAll('&quot;', '"')
      .replaceAll('&#39;', "'");
  return s.trim();
}

File: tv_apple/NasCabTV/Features/MediaLibrary/WebVttParser.swift
import Foundation

struct SubtitleCue {
    let startMs: Int64
    let endMs: Int64
    let text: String
}

enum SubtitleBitmapUtil {
    static func isBitmapCodecName(_ codecName: String?) -> Bool {
        let v = codecName?.trimmingCharacters(in: .whitespacesAndNewlines).lowercased() ?? ""
        return v == "pgssub"
            || v == "hdmv_pgs_subtitle"
            || v == "vobsub"
            || v == "dvd_subtitle"
            || v == "dvdsub"
            || v == "dvb_subtitle"
            || v == "xsub"
    }

    static func isBitmapExternalExtension(_ ext: String) -> Bool {
        let v = ext.trimmingCharacters(in: .whitespacesAndNewlines).lowercased()
        return v == ".sup" || v == ".sub" || v == ".idx"
    }
}

enum WebVttParser {
    static func parseWebVtt(_ vttText: String) -> [SubtitleCue] {
        let lines = vttText.replacingOccurrences(of: "\r\n", with: "\n").replacingOccurrences(of: "\r", with: "\n").split(separator: "\n", omittingEmptySubsequences: false).map(String.init)
        var cues: [SubtitleCue] = []
        var i = 0

        while i < lines.count, lines[i].trimmingCharacters(in: .whitespacesAndNewlines).isEmpty { i += 1 }
        if i < lines.count, lines[i].trimmingCharacters(in: .whitespacesAndNewlines).uppercased().hasPrefix("WEBVTT") {
            i += 1
        }

        while i < lines.count {
            while i < lines.count, lines[i].trimmingCharacters(in: .whitespacesAndNewlines).isEmpty { i += 1 }
            if i >= lines.count { break }

            let head = lines[i].trimmingCharacters(in: .whitespacesAndNewlines)
            if head.hasPrefix("NOTE") || head.hasPrefix("STYLE") || head.hasPrefix("REGION") {
                i += 1
                while i < lines.count, !lines[i].trimmingCharacters(in: .whitespacesAndNewlines).isEmpty { i += 1 }
                continue
            }

            var timeLine = lines[i]
            if !timeLine.contains("-->") {
                i += 1
                if i >= lines.count { break }
                timeLine = lines[i]
            }

            guard let arrowRange = timeLine.range(of: "-->") else {
                i += 1
                continue
            }
            let left = String(timeLine[..<arrowRange.lowerBound]).trimmingCharacters(in: .whitespacesAndNewlines)
            let rightPart = String(timeLine[arrowRange.upperBound...]).trimmingCharacters(in: .whitespacesAndNewlines)
            let right = rightPart.split(whereSeparator: \.isWhitespace).first.map(String.init)?.trimmingCharacters(in: .whitespacesAndNewlines) ?? ""

            guard let start = parseVttTimestamp(left), let end = parseVttTimestamp(right) else {
                i += 1
                while i < lines.count, !lines[i].trimmingCharacters(in: .whitespacesAndNewlines).isEmpty { i += 1 }
                continue
            }
            i += 1

            var textLines: [String] = []
            while i < lines.count, !lines[i].trimmingCharacters(in: .whitespacesAndNewlines).isEmpty {
                textLines.append(lines[i])
                i += 1
            }
            let text = formatCueText(textLines.joined(separator: "\n").trimmingCharacters(in: .whitespacesAndNewlines))
            if !text.isEmpty, end > start {
                cues.append(SubtitleCue(startMs: start, endMs: end, text: text))
            }
        }

        return cues.sorted { $0.startMs < $1.startMs }
    }

    static func findActiveCue(_ cues: [SubtitleCue], positionMs: Int64) -> SubtitleCue? {
        guard !cues.isEmpty else { return nil }
        var lo = 0
        var hi = cues.count - 1
        var best = -1
        while lo <= hi {
            let mid = (lo + hi) >> 1
            let c = cues[mid]
            if c.startMs <= positionMs {
                best = mid
                lo = mid + 1
            } else {
                hi = mid - 1
            }
        }
        if best < 0 { return nil }
        let c = cues[best]
        return positionMs >= c.startMs && positionMs <= c.endMs ? c : nil
    }

    private static func parseVttTimestamp(_ raw: String) -> Int64? {
        let s = raw.trimmingCharacters(in: .whitespacesAndNewlines)
        if s.isEmpty { return nil }
        let parts = s.split(separator: ":", omittingEmptySubsequences: false).map(String.init)
        guard parts.count >= 2, parts.count <= 3 else { return nil }

        let hours: Int
        let minutes: Int
        let seconds: Double
        if parts.count == 3 {
            guard let h = Int(parts[0]), let m = Int(parts[1]) else { return nil }
            hours = h
            minutes = m
            seconds = Double(parts[2].replacingOccurrences(of: ",", with: ".")) ?? -1
        } else {
            hours = 0
            guard let m = Int(parts[0]) else { return nil }
            minutes = m
            seconds = Double(parts[1].replacingOccurrences(of: ",", with: ".")) ?? -1
        }
        guard hours >= 0, minutes >= 0, seconds >= 0 else { return nil }
        return Int64(hours) * 3_600_000 + Int64(minutes) * 60_000 + Int64(seconds * 1000.0)
    }

    /// ffmpeg WebVTT 常带 `<br>`、`<font>` 等标签；叠层需转成纯文本。
    static func formatCueText(_ raw: String) -> String {
        var s = raw.trimmingCharacters(in: .whitespacesAndNewlines)
        if s.isEmpty { return s }
        if let br = try? NSRegularExpression(pattern: "<br\\s*/?>", options: [.caseInsensitive]) {
            let range = NSRange(s.startIndex..., in: s)
            s = br.stringByReplacingMatches(in: s, options: [], range: range, withTemplate: "\n")
        }
        if let tags = try? NSRegularExpression(pattern: "<[^>]*>") {
            let range = NSRange(s.startIndex..., in: s)
            s = tags.stringByReplacingMatches(in: s, options: [], range: range, withTemplate: "")
        }
        if let ass = try? NSRegularExpression(pattern: "\\{[^}]*\\}") {
            let range = NSRange(s.startIndex..., in: s)
            s = ass.stringByReplacingMatches(in: s, options: [], range: range, withTemplate: "")
        }
        s = s
            .replacingOccurrences(of: "&nbsp;", with: " ")
            .replacingOccurrences(of: "&amp;", with: "&")
            .replacingOccurrences(of: "&lt;", with: "<")
            .replacingOccurrences(of: "&gt;", with: ">")
            .replacingOccurrences(of: "&quot;", with: "\"")
            .replacingOccurrences(of: "&#39;", with: "'")
        return s.trimmingCharacters(in: .whitespacesAndNewlines)
    }
}
